### PR TITLE
preliminary integration of webhooks validator

### PR DESCRIPTION
**Pull request by mtrinh11**

### CODE REVIEW

**Comment by mtrinh11** [comment]
fixes https://github.com/xmidt-org/ancla/issues/14

File: tr1d1um.yaml [hunk 1]
```diff
@@ -123,6 +123,46 @@ log:
 # information using Argus.
 # Optional: if key is not supplied, webhooks would be disabled.
 webhook:
+
+  # validation provides options for validating the webhook's URL and TTL.
```
Comment (joe94) [inline tr1d1um.yaml:127]: We should add information on which fields are required to be added in the config file and if not what the default values are

File: deploy/packaging/tr1d1um_spruce.yaml [hunk 2]
```diff
@@ -122,6 +122,44 @@ log:
 # webhookStore provides configuration for storing and obtaining webhook
 # information using argus.
 webhook:
+  # validation provides options for validating the webhook's URL and TTL.
```
Comment (kristinapathak) [inline deploy/packaging/tr1d1um_spruce.yaml:125]:
```suggestion
  # validation provides options for validating the webhook's URL and TTL related fields.  URL fields that will be validated are __, __, and __; TTL related fields are Until and Duration.
```

File: deploy/packaging/tr1d1um_spruce.yaml [hunk 3]
```diff
@@ -122,6 +122,53 @@ log:
...
+    url:
+      # httpsOnly will allow only URLs with https schemes through if true.
+      # (Optional). Defaults to 'false'.
+      httpsOnly: false
+
+      # allowLoopback will disallow any canonical or IP loopback address if false.
+      # (Optional). Defaults to 'false'.
+      allowLoopback: true
+
+      # allowIP will disallow any IP addresses through if false.
```
Comment (joe94) [inline deploy/packaging/tr1d1um_spruce.yaml:139]:
```suggestion
      # allowIP allows the different webhook URLs to have IP hostnames if set to true.
```

File: deploy/packaging/tr1d1um_spruce.yaml [hunk 4]
```diff
@@ -122,6 +122,53 @@ log:
...
+      # (Optional). Defaults to 'false'.
+      allowIP: true
+
+      # allowSpecialUseHosts, if false, will disallow URLs that contain any string in invalidHosts or any string
+      # in SpecialUseHosts in webhookValidationConfig.go .
+      # (Optional). Defaults to 'false'.
```
Comment (joe94) [inline deploy/packaging/tr1d1um_spruce.yaml:145]:
```suggestion
      # allowSpecialUseHosts allows the different webhook URLs to have hosts specified by either config value invalidHosts or the SpecialUseHosts list in webhookValidationConfig.go if set to true.
      # (Optional). Defaults to 'false'.
```

File: deploy/packaging/tr1d1um_spruce.yaml [hunk 5]
```diff
@@ -122,6 +122,53 @@ log:
...
+      allowSpecialUseHosts: true
+
+      # allowSpecialUseIPs, if false, will disallow URLs that contain any string in invalidSubnets
+      # or any string in SpecialUseIPs in webhookValidationConfig.go .
+      # (Optional). Defaults to 'false'.
```
Comment (joe94) [inline deploy/packaging/tr1d1um_spruce.yaml:150]:
```suggestion
      # allowSpecialUseIPs allows the different webhook URLs to have IP hostnames that match any of the subnet masks specified by config value 'invalidSubnets' or specific IP addresses listed in SpecialUserIPs in webhookValidationConfig.go.
      # (Optional). Defaults to 'false'.
```

File: deploy/packaging/tr1d1um_spruce.yaml [hunk 6]
```diff
@@ -122,6 +122,53 @@ log:
...
+      allowSpecialUseIPs: true
+
+      # invalidHosts is a slice that contains strings of hosts that we do not want allowed in URLs
+      # if allowSpecialUseHosts is false.
+      # (Optional). Defaults to an empty slice.
+      invalidHosts: []
+
+      # invalidSubnets is a slice of strings of IPs that will be disallows in URLs if allowSpecialUseIPs 
+      # is false.
```
Comment (joe94) [inline deploy/packaging/tr1d1um_spruce.yaml:159]:
```suggestion
      # invalidSubnets is a list of subnet masks that will help capture a range of IP addresses that will be disallowed from webhook URLs when allowSpecialUseIPs is set to false.
```

**Review by joe94 — COMMENTED** [review]
Left some suggestions that I _think_ will make it easier to understand what the config values do. If they look good to you, please copy them to the main tr1d1um.yaml file 🎉

File: tr1d1um.yaml [hunk 7]
```diff
@@ -123,6 +123,53 @@ log:
...
+      # (Optional). Defaults to an empty slice.
+      invalidSubnets: []
+
+    # ttl is the webhook's time to live.
+    ttl:
+      # max is the length of time a webhook is allowed to live.
+      max: 1m
+
+      # jitter is the amount of additional time given to the webhook's TTL to account for
+      # any variations in delivery.
```
Comment (kristinapathak) [inline tr1d1um.yaml:170]:
```suggestion
      # jitter is the buffer time added when checking that the Until value is valid.  If there is a slight clock skew between servers or some delay in the http request, jitter should help account for that when ensuring that Until is not a time too far in the future.
```
The original comment makes it sound like we modify the webhook.Until value with the jitter before storing the webhook in argus.  My suggestion makes it more clear that jitter is only being used for validation purposes.

**Review by kristinapathak — APPROVED** [review]
looks great! 💯

**Review by kristinapathak — APPROVED** [review]

**Pull request merged**